src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,875 @@ pub(super) fn persist_sink<S>(
+        desired_frontiers: OkErr<Antichain<Timestamp>, Antichain<Timestamp>>,
+        /// The frontiers of the `persist` inputs.
+        persist_frontiers: OkErr<Antichain<Timestamp>, Antichain<Timestamp>>,
+        /// The current valid batch description and associated output capability, if any.
+        batch_description: Option<(BatchDescription, Capability<Timestamp>)>,
+    }
+
+    impl State {
+        fn new(
+            worker_id: usize,
+            persist_writer: WriteHandle<SourceData, (), Timestamp, Diff>,
+            metrics: SinkMetrics,
+        ) -> Self {
+            let worker_metrics = metrics.for_worker(worker_id);
+
+            Self {
+                worker_id,
+                persist_writer,
+                corrections: OkErr::new(
+                    Correction::new(metrics.clone(), worker_metrics.clone()),
+                    Correction::new(metrics, worker_metrics),
+                ),
+                desired_frontiers: OkErr::new_frontiers(),
+                persist_frontiers: OkErr::new_frontiers(),
+                batch_description: None,
+            }
+        }
+
+        fn advance_desired_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            advance(&mut self.desired_frontiers.ok, frontier);
+        }
+
+        fn advance_desired_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            advance(&mut self.desired_frontiers.err, frontier);
+        }
+
+        fn advance_persist_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.ok, frontier) {
+                self.apply_persist_frontier_advancement();
+            }
+        }
+
+        fn advance_persist_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.err, frontier) {
+                self.apply_persist_frontier_advancement();
+            }
+        }
+
+        /// Apply the effects of a previous `persist` frontier advancement.
+        fn apply_persist_frontier_advancement(&mut self) {
+            let frontier = self.persist_frontiers.frontier();
+
+            // We will only emit times at or after the `persist` frontier, so now is a good time to
+            // advance the times of stashed updates.
+            self.corrections.ok.advance_since(frontier.clone());
+            self.corrections.err.advance_since(frontier.clone());
+
+            // If the `persist` frontier is beyond the `lower` of the current batch description, we
+            // won't be able to append the batch, so the batch description is not valid anymore.
+            if let Some((desc, _)) = &self.batch_description {
+                if PartialOrder::less_than(&desc.lower, frontier) {
+                    self.batch_description = None;
+                }
+            }
+        }
+
+        fn absorb_batch_description(&mut self, desc: BatchDescription, cap: Capability<Timestamp>) {
```

Comment:
> Couldn't we get something like `[3, 5)` before `[1, 3)` or `[3, 5)` before `[1, 5)` if we got unlucky in the scheduling?

I'm not enough of a Timely expert to know for sure, but the operator assumes that this might happen and should handle this correctly. In your example, when an operator sees the `[3, 5)` batch description, it knows that any description with a smaller `lower` is not valid anymore (the `lower` of minted descriptions follows the persist frontier and we cannot append batches with a `lower` less than the current persist frontier). Thus when it later sees `[1, 3)` it should just discard it as an invalid batch description. Both `write` and `append` do that, in the "skipping outdated batch description" branches.

Do you see anything that relies on the order of batch descriptions, or batches?

Reply:
Sorry, how does the "skipping outdated batch description" branch do that here? It's comparing the lower of the incoming batch desc to `self.persist_frontiers` but not to anything in `self.batch_description`

Reply:
Good catch! I was assuming that `persist_frontier` is always up-to-data, i.e. that the `write` operator sees a `persist` advancement at the same time as the `mint` operator. But that isn't guaranteed and the `persist_frontier` could lag here. Changed to also compare with `self.batch_description.lower`.